### PR TITLE
Store relation name in manifest's node and source objects

### LINKED ISSUE #2647: Store resolved node names in manifest
### Description

A [neat wrinkle](https://github.com/fishtown-analytics/dbt/issues/2641#issuecomment-663241006):

> Currently, dbt doesn't store what the database form of a node should look like as a string when it runs. That means that given an existing manifest, the best we can do with the existing manifest is to tell you what the stored node would have been, if you loaded it with the current project's config - in particular, quoting.
>
> That's probably too restrictive! Instead, we should add a field to compiled nodes that indicates what the node's "resolved name" is. I'm pretty sure this can be filled in at compile-time, with something like str(adapter.Relation.create_from(config, node)).

### What does this get us?

* Better for deferral (https://github.com/fishtown-analytics/dbt/issues/2527, https://github.com/fishtown-analytics/dbt/issues/2641)
* Better `dbt ls` output (https://github.com/fishtown-analytics/dbt/issues/2643)
* dbt-docs display of relation name (we currently get quoting wrong)
* Future work around identifying orphaned database relations (https://github.com/fishtown-analytics/dbt/issues/1135)

### PR DESCRIPTION
**Pull request by franloza**
Resolves #2647

### Description

As @beckjake mentioned in #2641 , the relation name can be obtained from the adapter at compile-time. It was only necessary to add `relation` as an optional member of `CompiledNode` class.

In the tests, I added `relation_name` as the list of required keys for a compile node and values for `relation_name` in a couple of test cases (Despite the fact that it was not necessary for the test, I wanted to show this new field explicitly).

In development, the `manifest.json` included the field `relation_name` in every node object. Example:

```json
"model.sample_project.base_customers": {
      "raw_sql": "....",
      "compiled": true,
      ....
      "relation_name": "\"postgres\".\"public\".\"base_customers\""
    },
```

### Checklist
 - [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [X] I have run this code in development and it appears to resolve the stated issue
 - [X] This PR includes tests, or tests are not required/relevant for this PR
 - [X] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

**Review by jtcohen6 — COMMENTED** [review]
Thanks for picking this up @franloza! Always cool when a community member preempts us in sprint work :)

It looks like there are some failing tests in `029_docs_generate_tests`. You'll want to add `relation_name` as an expected value [here](https://github.com/fishtown-analytics/dbt/blob/dev/kiyoshi-kuromiya/test/integration/029_docs_generate_tests/test_docs_generate.py#L1046). Some of the other discrepancies in those test failures may have to do with other artifact changes that we've merged into `dev/kiyoshia-kuromiya`.

**Comment by franloza** [comment]
> It looks like there are some failing tests in `029_docs_generate_tests`. I think you'll want to add `relation_name` as an expected node value [here](https://github.com/fishtown-analytics/dbt/blob/dev/kiyoshi-kuromiya/test/integration/029_docs_generate_tests/test_docs_generate.py#L1046). Some of the other discrepancies in those test failures may have to do with other artifact changes that we've merged into `dev/kiyoshia-kuromiya`.

Yes! You're right, I forgot to add this new value in the doc generation tests. I'm fixing these tests and discrepancies with `dev/kiyoshia-kuromiya`.

File: test/integration/029_docs_generate_tests/test_docs_generate.py [hunk 1]
```diff
@@ -2616,11 +2677,16 @@ def expected_run_results(self, quote_schema=True, quote_model=False,
         schema = self.unique_schema()
 
         # we are selecting from the seed, which is always in the default db
-        compiled_database = self.default_database
-        if self.adapter_type != 'snowflake':
-            compiled_database = self._quote(compiled_database)
+        quote_database = self.adapter_type != 'snowflake'
```
Comment (jtcohen6) [inline test/integration/029_docs_generate_tests/test_docs_generate.py:2680]: nice touch here

**Review by jtcohen6 — COMMENTED** [review]
Nice work on getting all tests passing, especially the quoting, which is the real end goal of this effort.

I'm not quite sold on having `relation_name` defined for _all_ node objects. In particular, it seems inconsistent for tests and for ephemeral models, which currently have `null` for `database`, `schema`, and `alias`. Is there a way we could `null` out the `relation_name` for those resources?

**Comment by franloza** [comment]
Hey @jtcohen6, thank you for your review!

> I'm not quite sold on having `relation_name` defined for _all_ node objects. 

I agree with you about this. As you say, it is misleading for ephemeral models and tests. At this point, I think there might be multiple ways of nulling the relation name for these types. Let me come up with one implementation and let's discuss it. Any feedback or hint about the implementation is welcome!

**Comment by franloza** [comment]
I have considered the following criteria in my implementation to set a value for the `relation_name` field: The resource type is a non-ephemeral model, a seed or a snapshot
```
        if (node.resource_type in NodeType.refable() and
                node.config.materialized != "ephemeral"):
            adapter = get_adapter(self.config)
            relation_cls = adapter.Relation
            relation_name = str(relation_cls.create_from(self.config, node))
```

In addition, I missed a snapshot in the docs generation tests to verify that the relation name was the expected one. This is why I have decided to include a snapshot example in the tests. @jtcohen6 what do you think about this approach?

File: core/dbt/compilation.py [hunk 2]
```diff
@@ -172,6 +172,15 @@ def add_ephemeral_prefix(self, name: str):
         relation_cls = adapter.Relation
         return relation_cls.add_ephemeral_prefix(name)
 
+    def _get_relation_name(self, node: ParsedNode):
+        relation_name = None
+        if (node.resource_type in NodeType.refable() and
+                node.config.materialized != "ephemeral"):
```
Comment (jtcohen6) [inline core/dbt/compilation.py:178]:
```suggestion
                not node.is_ephemeral_model
```
based on similar logic [here](https://github.com/fishtown-analytics/dbt/blob/74eec3bdbee1e5be68c800ac400c0e6eac879e64/core/dbt/adapters/base/impl.py#L313-L314)

Comment (jtcohen6) [reply core/dbt/compilation.py:178]: also:
- could you add a quick inline comment to explain this?
- `refable()` excludes `sources`. they're not our primary objective here, but I see benefit in storing a `relation_name` for them, too. what do you think?

**Review by jtcohen6 — COMMENTED** [review]
Looking good! Tiny comment on the implementation.

Thanks for all the work around tests, and I'm really happy you added a snapshot example. Just wondering about disabling for BQ + Redshift—what's the common incompatibility?

**Comment by franloza** [comment]
Thanks again for your review! 

> Just wondering about disabling for BQ + Redshift—what's the common incompatibility?
There is no incompatibility, we are testing documentation for snapshots in `test__bigquery__run_and_generate` and `test__redshift__run_and_generate`. I did not want to include a snapshot in `test__bigquery__complex_models` and `test__redshift__incremental_view` to not make the PR bigger, but I think is good to include them also for these two methods for consistency. Let me include the snapshot test for these two methods.

File: core/dbt/compilation.py (hunk 2 above)
Comment (franloza) [reply core/dbt/compilation.py]: Sure! I think that's a good idea. Let me explore how can I include the `relation_name` for sources.

**Comment by franloza** [comment]
Hey @jtcohen6, finally I included the `resource_name` field for the `sources` node in the manifest. I checked locally and it seems to work:
```
  "sources": {
    "source.sample_project.base.customers": {
      "fqn": [ "sample_project", "base", "base", "customers" ],
      "database": "postgres",
      "schema": "public",
      "unique_id": "source.sample_project.base.customers",
      ...
      "patch_path": null,
      "unrendered_config": {},
      "relation_name": "\"postgres\".\"public\".\"customers\""
    }
```
I also modified the test to accept this new field. I did not modify the condition `if node.resource_type in NodeType.refable()`  because I needed to add the `resource_name` in a different module (`core/dbt/parser/schemas.py`) to appear in the `sources` object.

Finally, I added the snapshot to the tests `test__bigquery__complex_models` and `test__redshift__incremental_view`. What do you think about this implementation?

File: core/dbt/compilation.py [hunk 3]
```diff
@@ -172,6 +172,15 @@ def add_ephemeral_prefix(self, name: str):
         relation_cls = adapter.Relation
         return relation_cls.add_ephemeral_prefix(name)
 
+    def _get_relation_name(self, node: ParsedNode):
+        relation_name = None
+        if (node.resource_type in NodeType.refable() and
+                not node.is_ephemeral_model):
```
Comment (jtcohen6) [inline core/dbt/compilation.py:178]: 👍

**Review by jtcohen6 — APPROVED** [review]
Great work @franloza! Thank you so much for the contribution

**Comment by franloza** [comment]
> Great work @franloza! Thank you so much for the contribution

Thanks for your review! Glad to contribute 👍

**Pull request merged**